### PR TITLE
fix(platform-browser): `DomEventsPlugin` should always be the last plugin to be called for `supports()`.

### LINKED ISSUE #37149: HammerModule does not work when imported before BrowserModule
# 🐞 bug report

### Affected Package

This issue affects: @angular/platform-browser

### Is this a regression?
This issue exists in version 9.0 and 9.1.7. The HammerModule did not exist before 9.0.

### Description

 A clear and concise description of the problem...
If the `HammerModule` is imported before `BrowserModule` and `BrowserAnimationsModule`, it seems to have no effect.

## 🔬 Minimal Reproduction
[stackblitz](https://stackblitz.com/edit/angular-ivy-udwm55?file=src%2Fapp%2Fapp.module.ts) - notice that panning in the box does not log anything if HammerModule is imported before BrowserModule. If the order is switched, the component logs as expected.
## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.1.6
Node: 12.16.3
OS: darwin x64

Angular: 9.1.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.6
@angular-devkit/build-angular     0.901.6
@angular-devkit/build-optimizer   0.901.6
@angular-devkit/build-webpack     0.901.6
@angular-devkit/core              9.1.6
@angular-devkit/schematics        9.1.6
@angular/cdk                      9.2.2
@angular/cli                      9.1.6
@angular/fire                     6.0.0
@angular/flex-layout              9.0.0-beta.29
@angular/material                 9.2.3
@ngtools/webpack                  9.1.6
@schematics/angular               9.1.6
@schematics/update                0.901.6
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.42.0

</code></pre>

**Anything else relevant?**
I was unable to find any documentation about this requirement when importing `HammerModule`, so I assume it is a bug. If it's expected, a simple documentation addition may be enough!

### PR DESCRIPTION
**Pull request by JeanMeche**
This fixes the issues when `BrowserModule` is not the first module imported.

I choose not to introduce a specific token for the `DomEventsPlugin` as this would likely be breaking. 

Fixes #37149 #37850

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

File: packages/platform-browser/src/dom/events/interface.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {NgZone} from '@angular/core';
+
+/**
+ * @publicApi
+ */
+export interface EventManagerInterface {
```
Comment (JeanMeche) [inline packages/platform-browser/src/dom/events/interface.ts:14]: Had to create this interface to solve the circularity between `EventManagerPlugin` <=> `EventManager`

**Review by waterplea — APPROVED** [review]

File: packages/platform-browser/src/dom/events/interface.ts (hunk 1 above)
Comment (alxhub) [reply packages/platform-browser/src/dom/events/interface.ts:14]: I believe you could use `import type` instead, and avoid needing to introduce a new interface.

File: packages/platform-browser/src/dom/events/event_manager.ts [hunk 2]
```diff
@@ -37,7 +40,15 @@ export class EventManager {
     plugins.forEach((plugin) => {
       plugin.manager = this;
     });
-    this._plugins = plugins.slice().reverse();
+
+    const otherPlugins = plugins.filter(p => !(p instanceof DomEventsPlugin));
```
Comment (alxhub) [inline packages/platform-browser/src/dom/events/event_manager.ts:44]: Do we care about the tree-shakability of `DomEventsPlugin` here? We could put a marker field on it and look for that, rather than reference it directly here.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/platform-browser/src/dom/events/event_manager.ts (hunk 2 above)
Comment (JeanMeche) [reply packages/platform-browser/src/dom/events/event_manager.ts:44]: `DomEventsPlugin` is provided with the BrowserModule, so it will never be tree-shaken out right  ?

Comment (JeanMeche) [reply packages/platform-browser/src/dom/events/event_manager.ts]: Yes, I get your point, the bundle golden shouldn't change. I fixed it 👍

File: packages/platform-browser/src/dom/events/interface.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import type {EventManager} from './event_manager';
+
+export abstract class EventManagerPlugin {
+  // TODO: remove this unused property
```
Comment (JeanMeche) [inline packages/platform-browser/src/dom/events/interface.ts:11]: I wasn't quite sure if that cleanup would be breaking but I'll probably handle it with #49969 once this one is merged 👍

File: packages/platform-browser/src/dom/events/event_manager.ts [hunk 4]
```diff
@@ -11,6 +11,9 @@ import {Inject, Injectable, InjectionToken, NgZone, ɵRuntimeError as RuntimeErr
 
 import {RuntimeErrorCode} from '../../errors';
 
+import type {DomEventsPlugin} from './dom_events';
+import {EventManagerPlugin} from './interface';
```
Comment (alxhub) [inline packages/platform-browser/src/dom/events/event_manager.ts:15]: Not that it matters for the change, but it looks like this could be `import type` as well?

**Review by alxhub — APPROVED** [review]

**Review by alxhub — DISMISSED** [review]
Looking at the g3 failures, it looks like g3 expects to import `EventManagerPlugin` from the `event_manager.ts` file. We could update g3 when we land this change, but I don't think there's a reason to separate `EventManagerPlugin` into a separate file anymore - how would you feel about keeping it in `event_manager.ts`?

**Comment by JeanMeche** [comment]
It makes sense, we're down to 3 changed files.

**Comment by JeanMeche** [comment]
The current change ended up breaking at least one app inside Google. Currently investigating a bit why.

**Comment by JeanMeche** [comment]
I reverted to using `instanceof DomEventPlugin` and moving `EventManagerPlugin` to a separated file. 
TLDR: Having a dedicated `isDomEventPlugin` didn't work once it when through the closure compiler optimisations. 

Running a TGP to make sure we're good to go.

**Comment by JeanMeche** [comment]
[TGP is green](https://fusion2.corp.google.com/presubmit/810794749/OCL:810794749:BASE:815123035:1759600697139:c26d804b)

**Comment by JeanMeche** [comment]
Removing the merge label, this change still needs an approval

**Review by thePunderWoman — APPROVED** [review]
LGTM

**Comment by thePunderWoman** [comment]
This PR was merged into the repository. The changes were merged into the following branches:

- main: 29c397f8839234e0f9065979d721e583bf48ebb1
- 20.3.x: 2ece42866d0ee8240e73ebcef79ba47378777368
